Documentation/ExtensionArchitecture/Extbase/Reference/FileUpload.rst

```diff
@@ -0,0 +1,502 @@
+location, or specifying additional custom validators (which can be done via the PHP-API as
+described below)
+
+..  _extbase_fileupload_attribute_configuration:
+File upload configuration with the `FileUpload` attribute
+---------------------------------------------------------
+
+File upload for a property of a domain model can be configured using the
+newly introduced :php:`\TYPO3\CMS\Extbase\Annotation\FileUpload` attribute.
+
+Example:
+
+..  literalinclude:: _FileUpload/_BlogExcerpt.php
+    :caption: EXT:my_extension/Classes/Domain/Model/Blog.php (example excerpt of an Extbase domain model)
+    :emphasize-lines: 13-21
+
+All configuration settings of the
+:php:`\TYPO3\CMS\Extbase\Mvc\Controller\FileUploadConfiguration` object can
+be defined using the :php-short:`\TYPO3\CMS\Extbase\Annotation\FileUpload`
+attribute. It is however not possible
+to add custom validators using the
+:php-short:`\TYPO3\CMS\Extbase\Annotation\FileUpload` attribute, which you
+can achieve with a manual configuration as shown below.
+
+The currently available configuration array keys are:
+
+*   `validation` (:php:`array` with keys `required`, `maxFiles`, `minFiles`,
+    `fileSize`, `allowedMimeTypes`, `imageDimensions`, see
+    :ref:`extbase_fileupload_attribute-validationkeys`)
+*   `uploadFolder` (:php:`string`, destination folder)
+*   `duplicationBehavior` (:php:`object`, behaviour when file exists)
+*   `addRandomSuffix` (:php:`bool`, suffixing files)
+*   `createUploadFolderIfNotExist` (:php:`bool`, whether to create missing
+    directories)
+
+It is also possible to use the :php-short:`\TYPO3\CMS\Extbase\Annotation\FileUpload` annotation to configure
+file upload properties, but it is recommended to use the
+:php-short:`\TYPO3\CMS\Extbase\Annotation\FileUpload` attribute due to better readability.
+
+..  _extbase_fileupload_attribute-manual-configuration:
+
+Manual file upload configuration
+--------------------------------
+
+A file upload configuration can also be created manually and should be
+done in the :php:`initialize*Action`.
+
+Example:
+
+..  code-block:: php
+    :caption: Excerpt of an Extbase controller class
+
+    public function initializeCreateAction(): void
+    {
+        $mimeTypeValidator = GeneralUtility::makeInstance(MimeTypeValidator::class);
+        $mimeTypeValidator->setOptions(['allowedMimeTypes' => ['image/jpeg']]);
+
+        $fileHandlingServiceConfiguration = $this->arguments->getArgument('myArgument')->getFileHandlingServiceConfiguration();
+        $fileHandlingServiceConfiguration->addFileUploadConfiguration(
+            (new FileUploadConfiguration('myPropertyName'))
+                ->setRequired()
+                ->addValidator($mimeTypeValidator)
+                ->setMaxFiles(1)
+                ->setUploadFolder('1:/user_upload/files/')
+        );
+
+        $this->arguments->getArgument('myArgument')->getPropertyMappingConfiguration()->skipProperties('myPropertyName');
```

Comment:
I don't understand that line. I expect `myArgument` as the `Blog` argument. And `myPropertyName` as the FileReference property. If you skip that property in PMC the file will not be mapped on to that property. So, why defining a special configuration for that property above?

Reply:
It is required to exclude the property from extbase property mapper, since else extbase property mapper will try to map the property either as FileReference, which will fail (no TypeConverter). Therefore the extbase file upload API does not use extbase property mapper at all and does manual mapping _after_ extbase property mapper is finished.

Reply:
IMO this is an edge case because manual configuration is not something we really advertise or advocate. But for completioness we should keep it here. Hope the explanation is alright for you, resolving this.